#! /bin/bash
#PBS -N RNA_alignment
#PBS -o /public/home/fudan_bio/anyu/RNA_seq/RNA_hisat.log
#PBS -j oe
#PBS -l nodes=1:ppn=1
#PBS -l walltime=24:00:00
#PBS -q default

#软件路径
hisat=/public/home/fudan_bio/anyu/miniconda3/envs/rnaseq/bin/hisat2
samtools=/public/home/fudan_bio/anyu/miniconda3/envs/rnaseq/bin/samtools

#工作路径及变量（每次需更改）
workdir=/public/home/fudan_bio/anyu/RNA_seq
datadir=${workdir}/result/clean
outdir=${workdir}/result/alignment
logfile=${workdir}/log/hisat.log
index=${workdir}/reference/hg38

for file in $(ls ${datadir}/*.fastq.gz);do
	sample=`basename $file .fastq.gz`
	id=${sample%%R1.fastq.gz}
	logtime=$(date "+%Y-%m-%d %H:%M:%S")
	echo "------------------------------" >> $logfile
	echo "LOG TIME: $logtime" >> $logfile
	echo "FILE NAME: $sample" >> $logfile
	echo "------------------------------" >> $logfile
	$hisat -p 6 -x {index} \
		-1 ${id}R1.fastq.gz \
		-2 ${id}R2.fastq.gz \
		-S ${outdir}/sam/${id}.sam \
		2 > $logfile
	$samtools view -bS ${outdir}/sam/${id}.sam > ${outdir}/sam/${id}.bam
	$samtools sort -@ 1 ${outdir}/sam/${id}.bam -o ${outdir}/bam/${id}.sorted.bam
	$samtools index -@ 1 ${outdir}/bam/${id}.sorted.bam
done


		
	echo "------------end---------------" >>${logfile}
done



