#!/bin/bash

#PBS -N run_bwa-markdup-bqsr
#PBS -j oe
#PBS -l nodes=1:ppn=10
#PBS -l walltime=120:00:00
#PBS -q default

# 软件执行路径
bwa=$HOME/software/anaconda/anaconda3/envs/wes-analysis/bin/bwa
samtools=$HOME/software/anaconda/anaconda3/envs/wes-analysis/bin/samtools
gatk=$HOME/software/anaconda/anaconda3/envs/wes-analysis/bin/gatk

# 工作路径
workdir=$HOME/ASD_prof.AN
outdir=${workdir}/results/20201215
gatk_bundle=$HOME/ASD_prof.AN/data/gatk_bundle

# 批量执行变量: $sample

# 其他变量
refgenome=$HOME/ASD_prof.AN/data/reference_genome/gatk-hg19/ucsc.hg19.fa

# 程序执行时间及日志
logtime=$(date "+%Y-%m-%d %H:%M:%S")
logfile=${workdir}/logs/20201215/run_bwa-markdup-bqsr_${sample}.log
echo "------------------------------" >> $logfile
echo "LOG TIME: $logtime" >> $logfile
echo "------------------------------" >> $logfile

## 执行内容
# bwa
if [ -r ${outdir}/trimmomatic_output/pairedReads/${sample}_R1_paired.fastq.gz ] && [ -r ${outdir}/trimmomatic_output/pairedReads/${sample}_R2_paired.fastq.gz ]; then
    $bwa mem -t 10 -R "@RG\tID:${sample}\tSM:${sample}\tPL:illumina\tLB:${sample}" $refgenome \
        ${outdir}/trimmomatic_output/pairedReads/${sample}_R1_paired.fastq.gz \
        ${outdir}/trimmomatic_output/pairedReads/${sample}_R2_paired.fastq.gz | \
    $samtools view -Sb - > ${outdir}/bwa_output/${sample}_raw.bam
else
    $bwa mem -t 10 -R "@RG\tID:${sample}\tSM:${sample}\tPL:illumina\tLB:${sample}" $refgenome \
        ${workdir}/data/ASD_WESdata_20201215/FASTQ/*/Sample_${sample}/*_R1.fastq.gz \
        ${workdir}/data/ASD_WESdata_20201215/FASTQ/*/Sample_${sample}/*_R2.fastq.gz | \
    $samtools view -Sb - > ${outdir}/bwa_output/${sample}_raw.bam
fi 2>>$logfile
$samtools sort -@ 4 -m 4G -O bam \
    -o ${outdir}/bwa_output/${sample}_sorted.bam \
    ${outdir}/bwa_output/${sample}_raw.bam 2>>$logfile
rm ${outdir}/bwa_output/${sample}_raw.bam 2>>$logfile

# mark duplicates
$gatk MarkDuplicates -I ${outdir}/bwa_output/${sample}_sorted.bam \
    -O ${outdir}/gatk_output/${sample}_sorted.markdup.bam \
    -M ${outdir}/gatk_output/${sample}_markdup.metrics.txt 2>>$logfile
rm ${outdir}/bwa_output/${sample}_sorted.bam 2>>$logfile
cd ${outdir}/gatk_output/ 2>>$logfile
$samtools index ${sample}_sorted.markdup.bam 2>>$logfile

# BQSR
$gatk BaseRecalibrator -R $refgenome \
    -I ${outdir}/gatk_output/${sample}_sorted.markdup.bam \
    --known-sites ${gatk_bundle}/hg19/1000G_phase1.indels.hg19.vcf \
    --known-sites ${gatk_bundle}/hg19/Mills_and_1000G_gold_standard.indels.hg19.vcf \
    --known-sites ${gatk_bundle}/hg19/dbsnp_138.hg19.vcf \
    -O ${outdir}/gatk_output/${sample}_recal.data.table 2>>$logfile
$gatk ApplyBQSR --bqsr-recal-file ${outdir}/gatk_output/${sample}_recal.data.table \
    -R $refgenome \
    -I ${outdir}/gatk_output/${sample}_sorted.markdup.bam \
    -O ${outdir}/bam_output/${sample}_sorted.markdup.BQSR.bam 2>>$logfile
rm ${outdir}/gatk_output/${sample}_sorted.markdup.bam 2>>$logfile
cd ${outdir}/bam_output/ 2>>$logfile
$samtools index ${sample}_sorted.markdup.BQSR.bam 2>>$logfile

## 执行日志结束
echo "" >> $logfile

# END
